### PR TITLE
#1: Step 1 (closes #1)

### LINKED ISSUE #1: Step 1
First working implementation of the game

## requirements

- User can input “0”, “1”, “2” and play a game of Rock (0) Paper (1) Scissors (2).
- Create a `Game` object with a `def play(): Unit` method which asks for user input and prints the outcome of the game, including the two moves (user’s move and computer’s move)
- The computer’s move is generate at random

## misc 

Relevant points from the solution

- we’re using string interpolation with `s``""` 
- we’re using Strings directly
- we’re printing the result directly from the implementation of `play`, which in turn returns `Unit`

### Solution
https://github.com/buildo/rps/compare/step-0...step-1

### PR DESCRIPTION
**Pull request by raasoft**
Closes #1

### tests performed

kapture 2018-07-10 at 15 40 07

### tests not performed (domain coverage)

Not tested if the user is writing in input values different from:
- "0"
- "1"
- "2"

### CODE REVIEW

File: src/main/scala/rps/Game.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package rps;
```
Comment (gabro) [inline src/main/scala/rps/Game.scala]: super minor, but in scala it's convention not to use semicolons.

File: src/main/scala/rps/Game.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      println("You Win! 😤")
+    } else {
+      println("You Lose! 🤩")
+    }
+
+  }
+
+  private def returnMoveIcon(move: String): String = {
+    if (move == "0") {
+      return "💎"
```
Comment (gabro) [inline src/main/scala/rps/Game.scala]: another convention is Scala is to never use `return`.

The last expression in a block gets evaluated and returned automatically.

File: src/main/scala/rps/Game.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    } else if (move == "1") {
+      return "📄"
+    } else if (move == "2") {
+      return "✂️"
+    }
+    return ""
```
Comment (gabro) [inline src/main/scala/rps/Game.scala]: 👍 for handling the default case (hence making this function *total*). We'll see how to do this in a more principled way in the next steps ;)

File: src/main/scala/rps/Main.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+#%%%%%/*******#%%%%#*********%%%*********(%%%%#********/%%%
+#%%%%%%%/*******#%%%%(*****%%%%%%%*****/%%%%#********///%%%
+#%%%%%%%%%(*******#%%%%#/%%%%%%%%%%%/(%%%%#*******/%%%//%%%
+#%%%%%%%%%%%(*******#%%%%%%%%%%%%%%%%%%%#********%%%%%%%%%%
+#%%%%%%%%%%%%%/******%%%%%%%%%%%%%%%%%%%*******%%%%%%%%%%%%
+#%%%%%%%%%%%%%%%(*/%%%%%%%%%%%%%%%%%%%%%%%/*/%%%%%%%%%%%%%%
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Comment (gabro) [inline src/main/scala/rps/Main.scala:39]: lol :D

File: src/main/scala/rps/Game.scala [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+object Game {
+  def play(): Unit = {
+
+    /* Ask for user input */
+    val userMove = readLine(
+      "What's your move?\n" +
+        "0: " + returnMoveIcon("0") + "\n" +
+        "1: " + returnMoveIcon("1") + "\n" +
+        "2: " + returnMoveIcon("2") + "\n\n" +
+        "> ")
```
Comment (gabro) [inline src/main/scala/rps/Game.scala]: This is fine, but try to rewrite this using:

- [string interpolation](https://docs.scala-lang.org/overviews/core/string-interpolation.html) (`s""`)
- multiline strings with `stripMargin` (https://alvinalexander.com/scala/how-to-create-multiline-strings-heredoc-in-scala-cookbook)

**Review by gabro — CHANGES_REQUESTED** [review]
👍 good start!
Just some minor things before we can merge

File: src/main/scala/rps/Game.scala (hunk 1 above)
Comment (raasoft) [reply src/main/scala/rps/Game.scala]: You did well! Thank you for pointing that out!

File: src/main/scala/rps/Game.scala (hunk 2 above)
Comment (raasoft) [reply src/main/scala/rps/Game.scala]: [I've read that](https://tpolecat.github.io/2014/05/09/return.html): 
`A return expression, when evaluated, abandons the current computation and returns to the caller of the method in which return appears.`

Is this the "true" reason behind that convention?

File: build.sbt [hunk 6]
```diff
@@ -3,5 +3,5 @@ lazy val root = project
   .settings(
     name := "app-rpsosa",
     version := "1.0",
-    scalaVersion := "2.9.1"
+    scalaVersion := "2.12.3"
```
Comment (gabro) [inline build.sbt:6]: Woops, I've missed this. You can use 2.12.6 which is the latest version, but 2.12.3 is just fine.

File: src/main/scala/rps/Game.scala [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
-        "1: " + returnMoveIcon("1") + "\n" +
-        "2: " + returnMoveIcon("2") + "\n\n" +
-        "> ")
+    val userMove = readLine(s"""
+    #What's your move?"
+    #0: ${returnMoveIcon("0")}
+    #1: ${returnMoveIcon("1")}
+    #2: ${returnMoveIcon("2")}
+    #
+    #> """.stripMargin('#'))
```
Comment (gabro) [inline src/main/scala/rps/Game.scala:13]: this is 👍 but there's no need to define a custom separator. The default one is `|` and it works like

```scala
val userMove = readLine(s"""
    |What's your move?"
    |0: ${returnMoveIcon("0")}
    |1: ${returnMoveIcon("1")}
    |2: ${returnMoveIcon("2")}
    |
    |> """.stripMargin)
```

which is a bit more readable :)

**Review by gabro — APPROVED** [review]
👍 you can decide whether to apply the two suggestions, but otherwise it's good

File: build.sbt (hunk 6 above)
Comment (raasoft) [reply build.sbt:6]: I'll do it in the next step!

File: src/main/scala/rps/Game.scala (hunk 7 above)
Comment (raasoft) [reply src/main/scala/rps/Game.scala:13]: Gotcha! I wanted to try a new one, but in the next step I'll change it!

**Pull request merged**